### PR TITLE
Route accumulator

### PR DESCRIPTION
**Pull request by NelsonVides**
I'd like all routing to pass the accumulator along, with whatever modifications are needed like stripping it or updating the routed element. This will help with tracing the routing mechanism from start to end, if at all times the accumulator is preserved. It will also for example give me at the caller site, information about the packet handler's handling, like what muclight or pubsub did with the package. Then it is up to the caller to inspect the accumulator.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/69502) / small_tests / 1ed18c6762dda14049a852f069422b886ee68fbd
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3240/69502/small_tests.24.0.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3240/69502/small_tests.24.0.4/small/ct_run.mongooseim@localhost.2021-09-02_09.11.40/lib.mongooseim.logs/run.2021-09-02_09.11.40/suite.log.html)

File: src/ejabberd_router.erl [hunk 1]
```diff
@@ -547,28 +547,28 @@ route(_From, _To, Acc, _Packet, []) ->
...
+        {OrigFrom, OrigTo, Acc1, OrigPacketFiltered} ->
+            try xmpp_router:call_route(M, OrigFrom, OrigTo, Acc1, OrigPacketFiltered) of
+                {done, Acc2} ->
+                    mongoose_acc:append(router, result, {done, M}, Acc2);
                 {From, To, NAcc1, Packet} ->
                     route(From, To, NAcc1, Packet, Tail)
                 catch Class:Reason:Stacktrace ->
                     ?LOG_WARNING(#{what => routing_failed,
-                                   router_module => M, acc =>Acc,
+                                   router_module => M, acc => Acc0,
```
Comment (vkatsuba) [inline src/ejabberd_router.erl:562]: Do we need log `Acc0` or maybe `Acc1`? I'm ask cuz `mongoose_acc:append(router, result, {error, {M, Reason}}, Acc1)`

File: src/ejabberd_s2s.erl [hunk 2]
```diff
@@ -262,7 +262,7 @@ code_change(_OldVsn, State, _Extra) ->
                To :: jid:jid(),
                Acc :: mongoose_acc:t(),
                Packet :: exml:element()) ->
-        done. % this is the 'last resort' router, it always returns 'done'.
+    {done, mongoose_acc:t()}. % this is the 'last resort' router, it always returns 'done'.
```
Comment (vkatsuba) [inline src/ejabberd_s2s.erl:265]: Do we need return `{done, mongoose_acc:t()}.`? Maybe make sense return `mongoose_acc:t()`?

File: src/muc_light/mod_muc_light_codec_modern.erl [hunk 3]
```diff
@@ -68,12 +68,16 @@ encode({#msg{} = Msg, AffUsers}, Sender, {RoomU, RoomS} = RoomUS, HandleFun, Acc
                   affiliation => Aff,
                   role => mod_muc_light_utils:light_aff_to_muc_role(Aff)},
     HostType = mod_muc_light_utils:acc_to_host_type(Acc),
-    #xmlel{ children = Children }
+    Packet1 = #xmlel{ children = Children }
                      = mongoose_hooks:filter_room_packet(HostType, MsgForArch, EventData),
+                     ct:pal("Value ~p~n", [Packet1]),
```
Comment (vkatsuba) [inline src/muc_light/mod_muc_light_codec_modern.erl:73]: Looks like this(`ct:pal`) should be removed

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/69553) / small_tests / 878f4650ed3caa1fc7bdfda87209547ca69b9c76
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3240/69553/small_tests.24.0.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3240/69553/small_tests.24.0.4/small/ct_run.mongooseim@localhost.2021-09-02_10.10.49/lib.mongooseim.logs/run.2021-09-02_10.10.50/suite.log.html)

File: src/ejabberd_s2s.erl (hunk 2 above)
Comment (NelsonVides) [reply src/ejabberd_s2s.erl:265]: To match with more efficiency, and without looking into the `mongoose_acc`. If we want to match on the return value of these functions and check for the successful case, pattern-matching on a tagged tuple is more efficient than into a map with a specific key inside, and also, we would be exposing the internals of such map.

Comment (vkatsuba) [reply src/ejabberd_s2s.erl:265]: But this function always return the same, this mean that `mongoose_acc:t()` will always returned, and also this mean that we can check on `mongoose_acc:t()` in places where we need it. I do not insist, just was confusing of returning the same result always.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/69612) / small_tests / adb25226095606bf8eea2b2c42e99a16d5709f7f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3240/69612/small_tests.24.0.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3240/69612/small_tests.24.0.4/small/ct_run.mongooseim@localhost.2021-09-02_11.25.58/lib.mongooseim.logs/run.2021-09-02_11.25.58/suite.log.html)

**Review by vkatsuba — APPROVED** [review]
LGTM.

File: src/ejabberd_s2s.erl (hunk 2 above)
Comment (vkatsuba) [reply src/ejabberd_s2s.erl:265]: Yep. Make sense.

File: src/mongoose_local_delivery.erl [hunk 4]
```diff
@@ -41,6 +46,6 @@ do_route(OrigFrom, OrigTo, OrigAcc, OrigPacket, Handler) ->
                 drop ->
                     mongoose_hooks:xmpp_stanza_dropped(Acc0, OrigFrom,
                                                        OrigTo, OrigPacket),
-                    ok
+                    Acc0
```
Comment (chrzaszcz) [inline src/mongoose_local_delivery.erl:49]: Nice! It was against the spec.

**Review by chrzaszcz — COMMENTED** [review]
The code looks good, I only see one possible issue: if a host type and server is set in Acc, the call to `ejabberd_router:route(From, To, Acc, Packet)` will result (inc case of a locally routed packet) in an Acc that has a different host type and server (the ones for the recipient). As a result, further processing of such Acc might e.g. trigger hooks for a wrong host type.

To make sure it's not an issue, one would have to check all calls to `ejabberd_router:route/4` when the result is not ignored and ensure that the resulting Acc is never used to run hooks for a host type etc.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/69678) / small_tests / 79dcea61f4a4e7f0d1c25ed2df6b72091b887220
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3240/69678/small_tests.24.0.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3240/69678/small_tests.24.0.4/small/ct_run.mongooseim@localhost.2021-09-02_13.41.33/lib.mongooseim.logs/run.2021-09-02_13.41.34/suite.log.html)

**Review by chrzaszcz — APPROVED** [review]
Looks good

**Pull request merged**